[tyxiangs] iP

tyxiangs opened the pull request:
**### Dave: Your Friendly Task Companion**

Dave takes the stress out of remembering all those little (and big) things you need to do. It's:

- entirely text-based — just type and go!
- easy to pick up — no complicated manuals, just you and dave.
- LIGHTNING FAST — you'll be up and running before you finish your coffee.

Here’s how it works:

1. [https://drive.google.com/file/d/12W0K2WtDul_14HIHhUcBoTah_qRpueM9/view?usp=drive_link](url)
2. Give it a click and download it.
3. Add your tasks like a pro.
4. Sit back, and let dave handle the rest for you. 😎

And guess what? It’s 100% FREE! So why wait?

Key Features:

Task Management:
- Mark tasks as done or undone.
- Delete tasks from your list when no longer needed.
- Search for tasks by keywords in their description using the find command.
- View all tasks in a neat list format.
- Add tasks with different types:
1. Todo: A simple task without any date or time.
2. Deadline: A task with a specified due date and time.
3. Event: A task that specifies both a start and end time.

Task Status:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use dave as a foundation for practicing Java task management concepts.

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

tyxiangs committed:
Level 0 - Rename Chatbot

tyxiangs committed:
Level 0 - Greet

tyxiangs committed:
Level 0 - Chatbot greeting statement

tyxiangs committed:
Level 0 - Chatbot exit statement

tyxiangs committed:
Level1 - Echo

tyxiangs committed:
Level 1 Echo

tyxiangs committed:
Level 2. Add, List

tyxiangs committed:
Level 3 List Modification

tyxiangs committed:
Level 3 - Marking system completed

tyxiangs committed:
Level 4 - Completed implementing deadline, toDo and events class

tyxiangs committed:
Level 5 - Exception created and implement try and catch statement to catch potential error

tyxiangs committed:
Level 6 - Modify my code and use array list instead of array

tyxiangs committed:
Level 6 - Delete case done implemented

tyxiangs committed:
TextUiTesting

tyxiangs committed:
Enum implementation done

tyxiangs committed:
Level 7 - implement file function in task file

tyxiangs committed:
Level 7 - Implement save file func to update the file whenever there are changes

tyxiangs committed:
Merge branch 'branch-Level-7'

tyxiangs committed:
Level 8 - Implement date and time feature

tyxiangs committed:
Level 7 - String statement modification

tyxiangs committed:
Merge branch 'branch-Level-8'

tyxiangs committed:
A-MoreOOP - create classes like commands and modify the classess

tyxiangs committed:
Merge branch 'branch-A-MoreOOP'

tyxiangs committed:
A-Packages-place classes into respective package and do necessary import

tyxiangs committed:
Merge branch 'add-gradle-support'

tyxiangs committed:
A-Packages-Dave class modification

tyxiangs committed:
A-Gradle -  set up gradle

tyxiangs committed:
A-JUnit - add JUnit Test for todo, deadline and event

tyxiangs committed:
A-JUnit - test file modification

tyxiangs committed:
Level-8 - Add loadFile feature

tyxiangs committed:
A-jar - Create jar file

Ponharish commented on src/main/java/dave/storage/Storage.java:
```diff
@@ -0,0 +1,133 @@
+
+            try {
+                switch (taskType) {
+                    case "T": // Todo
```
Maybe you want to modify the style for switch statements as
`switch() { `
`case "T" :
`

'case' must be aligned with 'switch'

Ponharish commented on src/main/java/dave/storage/Storage.java:
```diff
@@ -0,0 +1,133 @@
+                        String eventStr = parts[3].trim();
+                        String[] eventParts = eventStr.split(" - ");
+                        LocalDateTime fromDate = LocalDateTime.parse(eventParts[0], dateFormatter);
+                        Event event = new Event(description + " /from " + fromDate.format(formatter) + " /to " + eventParts[1]);
```
This line might be too long, you may consider extending it to the next line

Ponharish commented on src/main/java/dave/task/Event.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Override
+    public String write() {
```
Perhaps you may be specific on where to write
eg: `writeToFile()`

Ponharish commented on src/main/java/dave/storage/Storage.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    public static void saveFile(TaskList dataList) throws IOException
+    {
```
You may want to consider moving the curly braces to the previous line

Ponharish commented on src/main/java/dave/parser/Parser.java:
```diff
@@ -0,0 +1,70 @@
+package dave.parser;
+
+import dave.command.*;
```
Perhaps, you may want to consider explicitly listing out all the libraries imported

Ponharish commented on src/main/java/dave/exceptions/InvalidCommandException.java:
```diff
@@ -0,0 +1,8 @@
+package dave.exceptions;
```
I like you idea of having different kinds of exception classes to handle different situations

Ponharish reviewed (commented):
Overall, a really good job done 👍 
I just highlighted a few minor changes which could be reviewed.

tyxiangs committed:
Add JavaDoc comments to method

tyxiangs committed:
Add JavaDoc comments to main class

tyxiangs committed:
Merge branch 'branch-A-JavaDoc'

tyxiangs committed:
Apply coding standard to written code

tyxiangs committed:
Merge branch 'A-CodingStandard'

tyxiangs committed:
Implement find feature

Create new command : create find command and implement execute func that find tasks by keywords
Modify parser : modify parser so it can identify new find command
Modify taskList : modify taskList so it can find matching tasks from stored tasks

tyxiangs committed:
Merge branch 'branch-Level-9'

itsme-zeix commented on src/main/java/dave/command/FindCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+public class FindCommand extends Command{
+
+    public String keyword;
```
```suggestion
    private String keyword;
```
Fields should be kept private unless the class is a data class with no behaviour.

itsme-zeix commented on src/main/java/dave/command/MarkCommand.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Represents the command to mark a task as done in the task list.
+ * This command modifies the task's status to indicate completion.
+ */
```
Small nitpick: try number of blank lines consistent. Currently the files switch between 1 and 2 new lines between imports and the JavaDoc for the class.

itsme-zeix commented on src/main/java/dave/storage/Storage.java:
```diff
@@ -0,0 +1,160 @@
+public class Storage {
+    private static String filePath;
+
+    String horizontal = "__________________________________________________________";
```
```suggestion
    private String HORIZONTAL = "__________________________________________________________";
```
1. Constant names must be all uppercase using underscore to separate words (aka SCREAMING_SNAKE_CASE).
2. Good to add access modifier unless package private is intended.

itsme-zeix commented on src/main/java/dave/task/Deadline.java:
```diff
@@ -0,0 +1,101 @@
+    public LocalDate dueDate;
+
+    /** The due time of the task */
+    public LocalTime dueTime;
```
You could consider changing the access modifier to private as the fields are currently only being accessed by tests.

Perhaps you could consider using stubs to deal with the behaviour of the class/method/fields so that there isn't a need to access the fields in `Deadline` directly.  Alternatively, a getter/setter method can be used.

itsme-zeix commented on src/main/java/dave/task/Task.java:
```diff
@@ -0,0 +1,66 @@
+    public String description;
+
+    /** The status of whether the task is done or not */
+    public boolean isDone;
```
You can consider making these fields private in favour of a getter/setter method instead.

itsme-zeix commented on src/main/java/dave/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @return {@code true} if this command is an exit command; {@code false} otherwise.
+     */
+    public boolean isExit() {
```
Should ideally use a verb to describe methods.

itsme-zeix commented on src/test/java/DeadlineTest.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Test
+    public void testDeadlineCreation_invalidDate() {
```
Test method should be renamed to follow this 3 part format: `featureUnderTest_testScenario_expectedBehavior()`
e.g. `sortList_emptyList_exceptionThrown()`, `getMember_memberNotFound_nullReturned()`

itsme-zeix reviewed (commented):
LGTM, just a few nitpicks with regards to style.

tyxiangs committed:
CheckStyle

Install checkstyle plugin and scan all the code
Make necessart=y change to code so it comply with coding style

tyxiangs committed:
Merge branch 'branch-A-CheckStyle'

tyxiangs committed:
GUI Design

Do GUI design
Modify my existing code to integrate gui feature inside

tyxiangs committed:
code modification

edit the code so it comply with coding standard

tyxiangs committed:
Merge branch 'branch-Level-10'

tyxiangs committed:
feat: Add assert feature and improve error handling

Added assertions to ensure TaskList, Storage, and Ui are not null before command execution to ensure component integrity. This enhances the reliability of the system by catching issues early in the code execution.

Refactored individual command classes (MarkCommand, UnmarkCommand, DeleteCommand, AddCommand, etc.) to return appropriate messages rather than printing directly to the console. This standardizes user feedback across the system.

Improved error handling by adding clearer messaging and ensuring proper responses are returned for invalid inputs. The code is now more resilient to edge cases, making the application more user-friendly.

Introduced assertComponents method in the Command class to centralize null checks, reducing code duplication. Additionally, added Javadoc comments for methods and classes to improve code documentation.

tyxiangs committed:
Refactor Parser to adhere to SLAP and reduce method complexity

Split the parse method in the Parser class into smaller, more focused methods to follow the Single Level of Abstraction Principle (SLAP) and to avoid long methods.

Extracted the splitCommand method to handle the parsing of the input command string, improving readability.
Moved the task type determination logic into a dedicated getTaskType method to isolate its responsibility.
Created separate methods, handleNumberCommand, handleDeadline, and handleEvent, for task-specific command handling, reducing the complexity in the main parse method.
Improved overall maintainability and readability by ensuring the parse method focuses only on high-level logic.
These changes simplify the codebase, making it easier to understand, maintain, and extend in the future.

tyxiangs committed:
Refactor Storage class and improve GUI documentation

Refactor Storage class methods to improve readability and follow SLAP principles.
Break down long methods into smaller, more manageable helper functions for file handling, task parsing, and error handling.

Main changes:

Split the loadFile method into smaller methods to handle different aspects of loading (e.g., reading tasks, parsing task data).
Enhanced separation of concerns by isolating file I/O and task-specific operations.
Improve GUI code clarity by adding Javadoc comments for MainWindow class methods.

tyxiangs committed:
Integrate Java Streams for task loading in Storage class

Refactored the Storage class to utilize Java Streams for improved efficiency and readability.

Replaced the traditional loop with Java Streams in Storage.loadFile() to handle the loading of tasks.
Used Collectors.toCollection(ArrayList::new) to ensure tasks are stored in an ArrayList.
This change optimizes the handling of task collections and leverages the power of Java Streams for cleaner and more concise code.

tyxiangs committed:
Merge pull request #1 from tyxiangs/branch-A-Assertions

feat: Add assert feature and improve error handling

tyxiangs committed:
Merge branch 'master' of https://github.com/tyxiangs/ip

tyxiangs committed:
Merge branch 'master' into branch-A-CodeQuality

tyxiangs committed:
Merge pull request #2 from tyxiangs/branch-A-CodeQuality

Branch a code quality

tyxiangs committed:
Merge branch 'master' into branch-A-Streams

tyxiangs committed:
Merge pull request #3 from tyxiangs/branch-A-Streams

Integrate Java Streams for task loading in Storage class

tyxiangs committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/MainWindow.java
#	src/main/java/dave/parser/Parser.java
#	src/main/java/dave/storage/Storage.java

tyxiangs committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/tyxiangs/ip into branch-A-CodeQuality

# Conflicts:
#	src/main/java/dave/parser/Parser.java
#	src/main/java/dave/storage/Storage.java

tyxiangs committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/dave/task/TaskList.java

tyxiangs committed:
Merge branch 'branch-A-Streams' of https://github.com/tyxiangs/ip into branch-A-Streams

tyxiangs committed:
Merge branch 'master' of https://github.com/tyxiangs/ip

tyxiangs committed:
Implement reminder feature as a command

- Added a `ReminderCommand` that allows users to type `reminder` to check all undone tasks and undone overdue tasks
- Added support for the `reminder overdue` command to check all undone tasks that are overdue.
- Both commands retrieve tasks from the task list and display them in a user-friendly format.
- Included a check to inform users when no undone or overdue tasks are found.

tyxiangs committed:
Merge branch 'branch-BCD-Extension'

tyxiangs committed:
Fix UI layout to properly display all dialog boxes with scrolling

- Removed fixed height constraint (`prefHeight`) from `VBox` (`dialogContainer`) to allow dynamic resizing as more messages are added.
- Enabled vertical scrollbar in `ScrollPane` by setting `vbarPolicy="ALWAYS"` to handle content overflow when dialog boxes exceed the visible area.
- Ensured `VBox` dynamically grows with its content inside the `ScrollPane` to prevent dialog boxes from being compressed or truncated.
- Adjusted `ScrollPane` and layout properties to support proper scrolling behavior for long messages.

tyxiangs committed:
Merge branch 'branch-A-BetterGui'

tyxiangs committed:
Refactor code for Storage, MainWindow, Dave, and Launcher classes:
- Added Javadoc comments for improved documentation
- Added assertions to ensure proper initialization and valid input
- Enhanced error handling with clearer messages
- Improved method decomposition for better adherence to SLAP
- Removed unused imports and ensured code cleanliness

tyxiangs committed:
Merge branch 'branch-AiAssisted'

tyxiangs committed:
Bug fixing

tyxiangs committed:
docs: Add user guide and reminder feature documentation

The user guide currently does not exist, and users may find it hard to
understand Dave's commands and features.

A user guide is necessary to help users understand Dave's various
features, such as adding tasks, setting reminders, and managing the
task list. The reminder feature also needed to be documented to help
users track undone and overdue tasks effectively.

This commit includes the following changes:
* Add a user guide to README.md with examples for adding tasks,
  setting deadlines, todos, events, and marking/unmarking tasks.
* Document the reminder feature with examples for `reminder all` and
  `reminder overdue`.
* Upload `Ui.png` to the docs folder and link it in README.md to
  provide a visual representation of the user interface.

The changes ensure users have clear, accessible information on how to
use the chatbot, and a UI screenshot for easy reference.

tyxiangs committed:
Merge branch 'branch-A-UserGuide'

tyxiangs committed:
fix: Correct index handling in DeleteCommand to fix off-by-one bug

The task list currently starts indexing tasks from 1, while the internal
data structure uses zero-based indexing. This caused an off-by-one error
when deleting tasks.

This commit adjusts the `DeleteCommand` constructor to subtract 1 from
the provided index, ensuring the correct task is deleted. Assertions
are also added to validate that the task list, storage, and UI objects
are not null and that the index is valid.

The changes fix the bug, ensuring the correct task is deleted from the
list based on the user's input.

tyxiangs committed:
fix: Ensure file is created if it does not exist

Previously, the storage system would fail when the specified file path
did not exist. This issue has now been addressed by adding logic to
ensure that the file is properly created if it doesn't already exist.

- Added checks to create directories and files when missing.
- Ensured that all file operations handle cases where the file may not
  exist initially.

This fix resolves the bug related to task file creation and ensures
that the application can properly handle missing files.

tyxiangs committed:
Merge branch 'branch-A-CodeQuality'

tyxiangs committed:
Merge branch 'branch-A-Streams'

# Conflicts:
#	src/main/java/dave/task/TaskList.java

tyxiangs committed:
feat: Add Deadline class with due date parsing and tests

- Implement Deadline class to support tasks with deadlines.
- Parse date and time using "yyyy-MM-dd HHmm" format with error handling.
- Add test case for valid deadline creation to ensure correct date and time parsing.
- Ensure compatibility with Java 17 as required for project submission.

tyxiangs committed:
feat: Implement window closure with delay after 'bye' command

Previously, the application did not close when the user entered the
"bye" command. This commit implements functionality to close the
window after the user enters "bye."

Additionally, a delay has been added using `PauseTransition` to
ensure the "bye" message is displayed for 2 seconds before the window
closes, allowing the user to read the goodbye message.

- Implemented window closure in `handleUserInput`.
- Added `PauseTransition` with a 2-second delay before closing.

This change improves the user experience by ensuring the "bye" message
is visible before the application exits.

tyxiangs committed:
Update README.md

docs: Add user guide and implement reminder feature

The user guide currently does not exist and users might find it 
difficult to understand Dave's commands and features.

A user guide is necessary to help users understand how to use the 
various features of Dave, such as adding tasks, setting reminders, and 
managing the task list. The reminder feature also needed to be 
documented to ensure users know how to track undone and overdue tasks.

This commit does the following:
* Adds a user guide to the README.md with clear instructions and 
  examples for adding tasks, setting deadlines, todos, events, and 
  marking/unmarking tasks.
* Implements documentation for the reminder feature, with examples for 
  both `reminder all` and `reminder overdue`.
* Uploads `Ui.png` to the docs folder and links it to the README.md 
  for visual representation.

The changes ensure that users have clear, accessible information to 
operate the chatbot, with instructions in the README and a visual aid 
in the form of a screenshot.

tyxiangs committed:
docs(README): update documentation to reflect current program features

Updated the README to ensure it is consistent with the latest version of the program. Changes include:

- Added new instructions that match the current functionality.
- Updated usage examples and feature descriptions.

This ensures the documentation is up-to-date and aligned with the program's current state.
